components/file_system/src/rate_limiter.rs

```diff
@@ -1,133 +1,362 @@
         IORateLimiterStatistics {
-            read: Default::default(),
-            write: Default::default(),
+            read_bytes: Default::default(),
+            write_bytes: Default::default(),
         }
     }
 
-    pub fn add(&self, io_type: IOType, io_op: IOOp, len: usize) {
+    pub fn fetch(&self, io_type: IOType, io_op: IOOp) -> usize {
+        let io_type_idx = io_type as usize;
+        match io_op {
+            IOOp::Read => self.read_bytes[io_type_idx].load(Ordering::Relaxed),
+            IOOp::Write => self.write_bytes[io_type_idx].load(Ordering::Relaxed),
+        }
+    }
+
+    pub fn record(&self, io_type: IOType, io_op: IOOp, bytes: usize) {
+        let io_type_idx = io_type as usize;
         match io_op {
             IOOp::Read => {
-                self.read[io_type as usize].fetch_add(len, Ordering::Relaxed);
+                self.read_bytes[io_type_idx].fetch_add(bytes, Ordering::Relaxed);
             }
             IOOp::Write => {
-                self.write[io_type as usize].fetch_add(len, Ordering::Relaxed);
+                self.write_bytes[io_type_idx].fetch_add(bytes, Ordering::Relaxed);
             }
         }
     }
 
-    pub fn fetch(&self, io_type: IOType, io_op: IOOp) -> usize {
-        match io_op {
-            IOOp::Read => self.read[io_type as usize].load(Ordering::Relaxed),
-            IOOp::Write => self.write[io_type as usize].load(Ordering::Relaxed),
+    pub fn reset(&self) {
+        for i in 0..IOType::VARIANT_COUNT {
+            self.read_bytes[i].store(0, Ordering::Relaxed);
+            self.write_bytes[i].store(0, Ordering::Relaxed);
         }
     }
+}
+
+const DEFAULT_REFILL_PERIOD: Duration = Duration::from_millis(200);
 
+#[inline]
+fn calculate_ios_per_refill(ios_per_sec: usize, refill_period: Duration) -> usize {
+    (ios_per_sec as f64 * refill_period.as_secs_f64()) as usize
+}
+
+/// Limit total IO flow below provided threshold by throttling low-priority IOs.
+/// Rate limit is disabled when total IO threshold is set to zero.
+#[derive(Debug)]
+struct RawIORateLimiter {
+    ios_through: [CachePadded<AtomicUsize>; IOPriority::VARIANT_COUNT],
+    ios_per_sec: [CachePadded<AtomicUsize>; IOPriority::VARIANT_COUNT],
+    pending_ios: [CachePadded<AtomicUsize>; IOPriority::VARIANT_COUNT],
+    protected: RwLock<RawIORateLimiterProtected>,
+}
+
+#[derive(Debug)]
+struct RawIORateLimiterProtected {
+    last_refill_time: Instant,
+}
+
+macro_rules! sleep_impl {
+    ($duration:expr, "sync") => {
+        std::thread::sleep($duration)
+    };
+    ($duration:expr, "async") => {
+        tokio::time::delay_for($duration).await
+    };
+    ($duration:expr, "non-blocking") => {
+        return 0
+    };
+}
+
+macro_rules! request_impl {
+    ($self:expr, $priority:expr, $amount:expr, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        loop {
+            let cached_ios_per_refill = $self.ios_per_sec[priority_idx].load(Ordering::Relaxed);
```

Comment:
Now that we don't have complex logic for high priority IOs, I think it's a good idea. Since EBS operates on a much lower level, the performance should be better (e.g. latency being more predictable)